### PR TITLE
OpenXR: Fix android loader extension detection

### PR DESCRIPTION
**Pull request by rsjtdrjgfuzkfg**
The OpenXR loader is initialized before enumerating extensions. This causes a warning about the lack of XR_KHR_loader_init_android to always be displayed, even if the runtime supports that extension.

While the cause of this is the OpenXR spec being inconsistent, we should only warn about a missing extension if the extension is actually missing. <strike>This PR thus moves the warning to be triggered after the loader is initialized.</strike> *After review and clarifications form the OpenXR working group, this PR now includes an improved warning if no extensions are present and fixes some additional aspects of loader extension detection.*

*This PR is fixing a very minor regression of https://github.com/godotengine/godot/pull/69392*

### CODE REVIEW

**Comment by BastiaanOlij** [comment]
Hi Dirk,

I agree with this. It's a little confusing what the idea behind this is, I'll ask some questions in the WG so I can find out some more but my gut feeling is that `XR_KHR_ANDROID_CREATE_INSTANCE_EXTENSION_NAME` actually replaces `XR_KHR_LOADER_INIT_ANDROID_EXTENSION_NAME` for exactly this reason, but we're still dealing with devices that implement the older approach (such as the quest).

As such I wonder if the warning is overkill to begin with. We could make an exit clause that if `xrInitializeLoaderKHR` fails and `XR_KHR_ANDROID_CREATE_INSTANCE_EXTENSION_NAME` is not available, then we fail, but if `XR_KHR_ANDROID_CREATE_INSTANCE_EXTENSION_NAME` is available we just ignore the failure of `xrInitializeLoaderKHR`

**Comment by BastiaanOlij** [comment]
Actually Dirk, @m4gr3d might be able to answer this or we should just try and test on Quest, but I think we can change the order of things in `OpenXRAPI::initialize` so that we call `on_before_instance_created` after we call `load_supported_extensions`.
`xrEnumerateInstanceExtensionProperties` should be callable and provide information before `xrInitializeLoaderKHR` is called. I think that method is badly named because it simply provides information about the VM for use when our instance is created.

If this works then we can also reliably check for the extension.

**Comment by BastiaanOlij** [comment]
Ok, this is getting funky, it's the other way around. `XR_KHR_android_create_instance` is the deprecated one but it hasn't been updated in the specs yet so it's interesting that PICO relies on it. `xrInitializeLoaderKHR` must be called first and before just about everything else.

Note in https://registry.khronos.org/OpenXR/specs/1.0/html/xrspec.html#XR_KHR_loader_init
```
On some platforms, before loading can occur the loader must be initialized with platform-specific parameters. Unlike other extensions, the presence of this extension is signaled by a successful call to [xrGetInstanceProcAddr](https://registry.khronos.org/OpenXR/specs/1.0/html/xrspec.html#xrGetInstanceProcAddr) to retrieve the function pointer for [xrInitializeLoaderKHR](https://registry.khronos.org/OpenXR/specs/1.0/html/xrspec.html#xrInitializeLoaderKHR) using a null instance handle. If this extension is supported, its use may be required on some platforms and the use of the [xrInitializeLoaderKHR](https://registry.khronos.org/OpenXR/specs/1.0/html/xrspec.html#xrInitializeLoaderKHR) function must precede other OpenXR calls except [xrGetInstanceProcAddr](https://registry.khronos.org/OpenXR/specs/1.0/html/xrspec.html#xrGetInstanceProcAddr). This function exists as part of the loader library that the application is using.
```

So it is clear that it can't be checked based on the extension, we get that information too late. The correct way to check is to see if we can obtain a function pointer.

Basically if `xrInitializeLoaderKHR_ptr` is `nullptr` after we execute `EXT_INIT_XR_FUNC`, we should just silently exit `OpenXRAndroidExtension::on_before_instance_created`. Ergo:
```
void OpenXRAndroidExtension::on_before_instance_created() {
	EXT_INIT_XR_FUNC(xrInitializeLoaderKHR);
	if (xrInitializeLoaderKHR_ptr == nullptr) {
		return;
	}
	...
```

**Comment by rsjtdrjgfuzkfg** [comment]
@BastiaanOlij as I read the spec, checking `xrInitializeLoaderKHR` only verifies that `XR_KHR_loader_init` is present (and needs to be done anyway for *any* loader that is initialized), but not that `XR_KHR_loader_init_android` is present and enabled (and what does "enabled" even mean in this context?!). Imho the constraint "The [XR_KHR_loader_init_android] extension must be enabled prior to using XrLoaderInitInfoAndroidKHR" just does not really make sense.

But in practical matters: if you agree to remove the warning I can update the PR to do that. I just don't like keeping in a warning that is always displayed ;)

**Comment by BastiaanOlij** [comment]
Hey @rsjtdrjgfuzkfg ,

Sorry got a bit confusing there because my first two comments were written before I talked to the WG guys. So that last comment is the important one and especially the

> On some platforms, before loading can occur the loader must be initialized with platform-specific parameters. Unlike other extensions, the presence of this extension is signaled by a successful call to [xrGetInstanceProcAddr](https://registry.khronos.org/OpenXR/specs/1.0/html/xrspec.html#xrGetInstanceProcAddr) to retrieve the function pointer for [xrInitializeLoaderKHR](https://registry.khronos.org/OpenXR/specs/1.0/html/xrspec.html#xrInitializeLoaderKHR) using a null instance handle. If this extension is supported, its use may be required on some platforms and the use of the [xrInitializeLoaderKHR](https://registry.khronos.org/OpenXR/specs/1.0/html/xrspec.html#xrInitializeLoaderKHR) function must precede other OpenXR calls except [xrGetInstanceProcAddr](https://registry.khronos.org/OpenXR/specs/1.0/html/xrspec.html#xrGetInstanceProcAddr). This function exists as part of the loader library that the application is using.

So yes, bottom line is, remove the warning, there is no point there. In theory we could even remove the checking for the `XR_KHR_LOADER_INIT_ANDROID_EXTENSION_NAME` extension all together because in this particular case it does nothing.

All we need to do is add the `xrInitializeLoaderKHR_ptr == nullptr` check, if there is no pointer to this function we're dealing with a loader that doesn't have an initialiser.

If you look at the spec of the loader function itself, you'll see that there is a requirement of the XR runtime that if it implements this function, it MUST support the structure relevant to the platform, and only for Android such a structure is defined. So in practice this means `xrInitializeLoaderKHR` is only implemented on Android and will always have the Android structure as its input.

**Comment by rsjtdrjgfuzkfg** [comment]
@BastiaanOlij ok, so I think I now have updated the check to work properly. Note that a null pointer check is imho not sufficient, as our `INIT_XR_FUNC` macro family includes return statements and error logging that would interfere. I also added a warning if both mechanisms are unavailable, though I'm open to removing it if you feel it is overkill.

I'd still like to see the working group to remove the sentence "The [XR_KHR_loader_init_android] extension must be enabled prior to using XrLoaderInitInfoAndroidKHR" from the "Valid Usages (implicit)" section of [the documentation for `XrLoaderInitInfoAndroidKHR`](https://registry.khronos.org/OpenXR/specs/1.0/man/html/XrLoaderInitInfoAndroidKHR.html), but that is not related to godot. ;)

**Comment by BastiaanOlij** [comment]
> I'd still like the working group to remove the sentence "The [XR_KHR_loader_init_android] extension must be enabled prior to using XrLoaderInitInfoAndroidKHR" from the "Valid Usages (implicit)" section of [the documentation for `XrLoaderInitInfoAndroidKHR`](https://registry.khronos.org/OpenXR/specs/1.0/man/html/XrLoaderInitInfoAndroidKHR.html), but that is not related to godot. ;)

Ah! I was only looking at the specification for the call itself, totally missed you pointing that out. I've passed on the message :)

**Review by BastiaanOlij — APPROVED** [review]

**Comment by akien-mga** [comment]
Thanks!
